pallets/clusters/src/lib.rs

```diff
@@ -0,0 +1,479 @@
+	pub enum Event<T: Config> {
+		ClusterCreated {
+			cluster_hash: Hash128,
+		},
+		RoleCreated {
+			role_hash: Hash128,
+		},
+		RoleEdited {
+			role_hash: Hash128,
+		},
+		RoleDeleted {
+			role_hash: Hash128,
+		},
+		/// A new sponsored account was added
+		ProxyAccountAdded {
+			cluster_account: T::AccountId,
+			proxy_account: T::AccountId,
+		},
+	}
+
+	// Errors inform users that something went wrong.
+	#[pallet::error]
+	pub enum Error<T> {
+		/// The cluster already exists.
+		ClusterExists,
+		/// Cluster not found
+		ClusterNotFound,
+		/// The role in the cluster already exists.
+		RoleExists,
+		/// Element not found
+		RoleNotFound,
+		/// RoleSettings not found
+		RoleSettingsNotFound,
+		/// Missing permission to perform an operation
+		NoPermission,
+		/// Invalid inputs
+		InvalidInput,
+		/// Technical error not categorized
+		SystematicFailure,
+		/// The owner is not correct.
+		OwnerNotCorrect,
+		/// The member already exists in the cluster
+		MemberExists,
+		/// Too many members
+		MembersLimitReached,
+		/// Member not found in the cluster
+		MemberNotFound,
+		/// Account proxy already associated with the cluster account
+		AccountProxyAlreadyExists,
+		/// Too many proxies associated with the cluster
+		TooManyProxies,
+		/// InsufficientBalance
+		InsufficientFunds,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Create a **Cluster** passing a name as input.
+		#[pallet::weight(10_000 + T::DbWeight::get().writes(1))]
+		pub fn create_cluster(
+			origin: OriginFor<T>,
+			name: BoundedVec<u8, T::NameLimit>,
+		) -> DispatchResult {
+			let who = ensure_signed(origin.clone())?;
+
+			ensure!(!name.len().is_zero(), Error::<T>::InvalidInput);
+
+			// as_be_bytes to get bytes without allocations
+			let extrinsic_index = <frame_system::Pallet<T>>::extrinsic_index()
+				.ok_or(Error::<T>::SystematicFailure)?
+				.to_be_bytes();
+			// BlockNumber type is non concrete, so we need to encode it to get bytes
+			let current_block_number = <frame_system::Pallet<T>>::block_number().encode();
+			// Who is not concrete, so we need to encode it to get bytes
+			let who_bytes = who.encode();
+
+			let cluster_id = blake2_128(
+				&[&current_block_number, name.as_slice(), &extrinsic_index, &who_bytes].concat(),
+			);
+
+			// Check that the cluster does not exist already
+			ensure!(!<Clusters<T>>::contains_key(&cluster_id), Error::<T>::ClusterExists);
+
+			// At creation there are no roles and no members assigned to the cluster
+			let cluster = Cluster {
+				owner: who.clone(),
+				name: name.into_inner(), // `self.0`
+				cluster_id,
+				roles: vec![],
+				members: vec![],
+			};
+
+			let minimum_balance =
+				<pallet_balances::Pallet<T> as fungible::Inspect<T::AccountId>>::minimum_balance();
+			let origin_balance =
+				<pallet_balances::Pallet<T> as fungible::Inspect<T::AccountId>>::balance(&who);
+			ensure!(origin_balance > minimum_balance, Error::<T>::InsufficientFunds);
```

Comment:
why do we need to ensure that the user's balance is greater than the existential balance?

Reply:
it is not needed, indeed. I can remove it.